From 450f0bf62bade57f077d2ece410746eb7be73747 Mon Sep 17 00:00:00 2001
From: Manish Goregaokar <manishearth@google.com>
Date: Sat, 1 Nov 2025 18:40:29 +0000
Subject: [PATCH] Constrain durations proactively before passing to ICU4X

This uplifts https://github.com/boa-dev/temporal/pull/615. It ought to
prevent debug assertions and pathologically slow behavior when
attempting to do date arithmetic with very large durations. These
durations are too large to produce a valid in-range date anyway, so we
can proactively error when we see them.

Change-Id: Id2426776556b4451335b7d0f775b43b3c5b92ac7
---
 .../src/builtins/compiled/duration/tests.rs   | 21 +++++++++-
 .../src/builtins/core/calendar.rs             | 39 +++++++++++++++++++
 2 files changed, 59 insertions(+), 1 deletion(-)

diff --git a/third_party/rust/chromium_crates_io/vendor/temporal_rs-v0_1/src/builtins/compiled/duration/tests.rs b/third_party/rust/chromium_crates_io/vendor/temporal_rs-v0_1/src/builtins/compiled/duration/tests.rs
index 72ae0e4a1f67b..0b6acac85cb18 100644
--- a/third_party/rust/chromium_crates_io/vendor/temporal_rs-v0_1/src/builtins/compiled/duration/tests.rs
+++ b/third_party/rust/chromium_crates_io/vendor/temporal_rs-v0_1/src/builtins/compiled/duration/tests.rs
@@ -1,7 +1,8 @@
 use crate::{
     duration::DateDuration,
     options::{
-        OffsetDisambiguation, RelativeTo, RoundingIncrement, RoundingMode, RoundingOptions, Unit,
+        OffsetDisambiguation, Overflow, RelativeTo, RoundingIncrement, RoundingMode,
+        RoundingOptions, Unit,
     },
     partial::PartialDuration,
     Calendar, PlainDate, TimeZone, ZonedDateTime,
@@ -626,6 +627,24 @@ fn add_normalized_time_duration_out_of_range() {
     assert!(err.is_err())
 }
 
+#[test]
+fn add_large_durations() {
+    // Testcases found by fuzzing <https://github.com/unicode-org/icu4x/pull/7206>
+    let base = PlainDate::new(2000, 1, 1, Calendar::from_str("dangi").unwrap()).unwrap();
+
+    let test_duration = Duration::from(DateDuration::new(4294901760, 256, 0, 0).unwrap());
+    assert!(base.add(&test_duration, Some(Overflow::Constrain)).is_err());
+
+    let test_duration = Duration::from(DateDuration::new(0, 1281, 0, 8589934592).unwrap());
+    assert!(base.add(&test_duration, Some(Overflow::Constrain)).is_err());
+
+    let test_duration = Duration::from(DateDuration::new(2046820352, 0, 0, 0).unwrap());
+    assert!(base.add(&test_duration, Some(Overflow::Constrain)).is_err());
+
+    let test_duration = Duration::from(DateDuration::new(0, 0, 2516582400, 0).unwrap());
+    assert!(base.add(&test_duration, Some(Overflow::Constrain)).is_err());
+}
+
 #[test]
 fn test_rounding_boundaries() {
     let relative_to = PlainDate::new(2000, 1, 1, Calendar::default()).unwrap();
diff --git a/third_party/rust/chromium_crates_io/vendor/temporal_rs-v0_1/src/builtins/core/calendar.rs b/third_party/rust/chromium_crates_io/vendor/temporal_rs-v0_1/src/builtins/core/calendar.rs
index d1508eee29c8d..d4b55eb74529e 100644
--- a/third_party/rust/chromium_crates_io/vendor/temporal_rs-v0_1/src/builtins/core/calendar.rs
+++ b/third_party/rust/chromium_crates_io/vendor/temporal_rs-v0_1/src/builtins/core/calendar.rs
@@ -250,6 +250,43 @@ impl<'a> TryFrom<&'a CalendarFields> for DateFields<'a> {
         Ok(this)
     }
 }
+
+/// See <https://github.com/unicode-org/icu4x/issues/7207>
+///
+/// We don't want to trigger any pathological or panicky testcases in ICU4X.
+///
+/// To aid in that, we early constrain date durations to things that have no hope of producing a datetime
+/// that is within Temporal range.
+fn early_constrain_date_duration(duration: &IcuDateDuration) -> Result<(), TemporalError> {
+    // Temporal range is -271821-04-20 to +275760-09-13
+    // This is (roughly) the maximum year duration that can exist for ISO
+    const TEMPORAL_MAX_ISO_YEAR_DURATION: u32 = 275760 + 271821;
+    // Double it. No calendar has years that are half the size of ISO years.
+    const YEAR_DURATION: u32 = 2 * TEMPORAL_MAX_ISO_YEAR_DURATION;
+    // Assume every year is a leap year, calculate a month range
+    const MONTH_DURATION: u32 = YEAR_DURATION * 13;
+    // Our longest year is 390 days
+    const DAY_DURATION: u32 = YEAR_DURATION * 390;
+    const WEEK_DURATION: u32 = DAY_DURATION / 7;
+
+    let err = Err(TemporalError::range().with_enum(ErrorMessage::IntermediateDateTimeOutOfRange));
+
+    if duration.years > YEAR_DURATION {
+        return err;
+    }
+    if duration.months > MONTH_DURATION {
+        return err;
+    }
+    if duration.weeks > WEEK_DURATION {
+        return err;
+    }
+    if duration.days > DAY_DURATION.into() {
+        return err;
+    }
+
+    Ok(())
+}
+
 // ==== Public `CalendarSlot` methods ====
 
 impl Calendar {
@@ -437,6 +474,8 @@ impl Calendar {
             weeks: u32::try_from(duration.weeks.abs()).map_err(|_| invalid)?,
             days: u64::try_from(duration.days.abs()).map_err(|_| invalid)?,
         };
+
+        early_constrain_date_duration(&duration)?;
         let mut options = DateAddOptions::default();
         options.overflow = Some(overflow.into());
         let calendar_date = self.0.from_iso(*date.to_icu4x().inner());
-- 
2.51.1.930.gacf6e81ea2-goog

